### PR TITLE
Impl reservation entity

### PR DESCRIPTION
**Pull request by MaltsevaKsenia**
I will add unit tests later

### CODE REVIEW

File: src/main/java/com/epam/controller/MainController.java [hunk 1]
```diff
@@ -43,12 +37,12 @@ public String badRequest(Principal user, ModelMap modelMap) {
     }
 
     @GetMapping("/login")
-    public String loginPage (Principal user, ModelMap modelMap) {
+    public String loginPage(Principal user, ModelMap modelMap) {
```
Comment (ghost) [inline src/main/java/com/epam/controller/MainController.java]: please avoid some pointless changes

File: src/main/java/com/epam/controller/ReservationController.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+
+import java.security.Principal;
+import java.util.List;
+
+@Controller
+public class ReservationController {
+    @Autowired
+    ReservationService reservationService;
```
Comment (ghost) [inline src/main/java/com/epam/controller/ReservationController.java]: this variable can be private final and autowired using constructor, because it is easy t understand and write tests

File: src/main/java/com/epam/controller/ReservationController.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    @GetMapping("/testadmin")
+    public String listReservations(Principal user, ModelMap modelMap) {
+        List<Reservation> list = reservationService.listReservations();
+        modelMap.addAttribute("listReservation", list);
+        return "testadmin";
+    }
+
+    @RequestMapping(value = "/delete/{id}", method = RequestMethod.GET)
```
Comment (ghost) [inline src/main/java/com/epam/controller/ReservationController.java]: Why not to use Delete HTTP request method?

Comment (ghost) [reply src/main/java/com/epam/controller/ReservationController.java]: URLs of the controllers should be unique. Do you think, that just only 'delete' will be here?

File: src/main/java/com/epam/model/Reservation.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package com.epam.model;
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
```
Comment (ghost) [inline src/main/java/com/epam/model/Reservation.java:8]: I'm not sure, but no args/all args constructor is already defined in Data/Builder annotation?

File: src/main/java/com/epam/model/Reservation.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+@AllArgsConstructor
+@Builder
+@Data
+public class Reservation {
+    private Integer id;
+    private Integer client_id;
```
Comment (ghost) [inline src/main/java/com/epam/model/Reservation.java]: Oh, no no, please follow Java conventions about variables naming

File: src/main/java/com/epam/model/Reservation.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    private Integer tourOffer_id;
+    private Integer numberOfPeople;
+    private String status;
```
Comment (ghost) [inline src/main/java/com/epam/model/Reservation.java]: Maybe status can be enum?

File: src/main/java/com/epam/repository/PersonDAO.java [hunk 7]
```diff
@@ -26,7 +26,7 @@ public PersonDAO(NamedParameterJdbcTemplate jdbcTemplate) {
     Person buildPerson(ResultSet rs) throws SQLException {
         return Person.builder()
                 .id(rs.getInt("id"))
-                .name(rs.getString("nickname"))
+                .name(rs.getString("email"))
```
Comment (ghost) [inline src/main/java/com/epam/repository/PersonDAO.java]: How it will used in registration? It is security functionality... If you want to fix it - please make another pull request with this hot-fix. Because, imagine, that you will fix a lot of other logic, and how the reviewers will find your task logic?

File: src/main/java/com/epam/repository/ReservationDAO.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.stereotype.Repository;
+
+import javax.sql.DataSource;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.List;
+@NoArgsConstructor
```
Comment (ghost) [inline src/main/java/com/epam/repository/ReservationDAO.java]: Do we need it here?

File: src/main/java/com/epam/repository/ReservationDAO.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+@Repository
+public class ReservationDAO {
+    private JdbcTemplate jdbcTemplate;
```
Comment (ghost) [inline src/main/java/com/epam/repository/ReservationDAO.java]: This can be final

File: src/main/java/com/epam/config/DBConfiguration.java [hunk 10]
```diff
@@ -28,4 +29,9 @@ public DataSource dataSource() {
     public NamedParameterJdbcTemplate namedParamJdbcTemplate() {
         return new NamedParameterJdbcTemplate(dataSource());
     }
+
+    @Bean
```
Comment (ghost) [inline src/main/java/com/epam/config/DBConfiguration.java]: Do we need both jdbcTemplate and namedParamJdbcTemplate beans?

File: src/main/java/com/epam/repository/ReservationDAO.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+                .client_id(rs.getInt("client_id"))
+                .tourOffer_id(rs.getInt("tourOffer_id"))
+                .numberOfPeople(rs.getInt("numberOfPeople"))
+                .status(rs.getString("status"))
+                .discount_id(rs.getInt("discount_id"))
+                .totalPrice(rs.getInt("totalPrice"))
+                .build();
+    }
+
+    public void addReservation(Reservation reservation) {
```
Comment (ghost) [inline src/main/java/com/epam/repository/ReservationDAO.java]: How will we check the validity of the argument? (in all methods)

Comment (ghost) [reply src/main/java/com/epam/repository/ReservationDAO.java]: Is it possible to check, that adding was successful?

File: src/main/java/com/epam/repository/ReservationDAO.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+                reservation.getStatus(), reservation.getDiscount_id(), reservation.getTotalPrice());
+    }
+
+    public Reservation getReservationById(Integer id) {
+        String sql = "SELECT * FROM reservation WHERE id = ?";
+        return jdbcTemplate.queryForObject(sql,
+                new Object[]{id}, (rs, rowNum) -> buildReservation(rs));
+    }
+
+    public List listReservations() {
```
Comment (ghost) [inline src/main/java/com/epam/repository/ReservationDAO.java]: Oh no,no. Raw list, without generics????

File: src/main/java/com/epam/repository/ReservationDAO.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+        String sql = "SELECT * from reservation";
+        return jdbcTemplate.query(sql, reservationMapper);
+    }
+
+    public void removeReservation(Integer id) {
+        String sql = "DELETE FROM reservation WHERE id = ?";
+        jdbcTemplate.update(sql, id);
+    }
+
+    public void updateReservation(Integer id, Integer Client_id, Integer TourOffer_id, Integer NumberOfPeople, String Status, Integer Discount_id, Integer TotalPrice) {
```
Comment (ghost) [inline src/main/java/com/epam/repository/ReservationDAO.java]: You can use object here, that includes all your args

File: src/main/java/com/epam/repository/ReservationDAO.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+                new Object[]{id}, (rs, rowNum) -> buildReservation(rs));
+    }
+
+    public List listReservations() {
+        String sql = "SELECT * from reservation";
+        return jdbcTemplate.query(sql, reservationMapper);
+    }
+
+    public void removeReservation(Integer id) {
+        String sql = "DELETE FROM reservation WHERE id = ?";
```
Comment (ghost) [inline src/main/java/com/epam/repository/ReservationDAO.java]: IMHO, table names and column names can be stored in some place like property file or singleton like a static final strings. In future, when we will want to rename our columns - we will fix it in one place.

File: src/main/java/com/epam/service/ReservationService.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    private final ReservationDAO reservationDAO;
+
+
+    @Autowired
+    public ReservationService(ReservationDAO reservationDAO) {
+        this.reservationDAO = reservationDAO;
+    }
+
+
+    public void addReservation(Reservation reservation) {
```
Comment (ghost) [inline src/main/java/com/epam/service/ReservationService.java]: How do you want to check the validity of the arguments?

File: src/main/java/com/epam/repository/ReservationDAO.java (hunk 13 above)
Comment (ghost) [reply src/main/java/com/epam/repository/ReservationDAO.java]: Please, fix this in all places in your PR

File: src/main/java/com/epam/repository/ReservationDAO.java (hunk 12 above)
Comment (ghost) [reply src/main/java/com/epam/repository/ReservationDAO.java]: Please, fix it everywhere

File: src/main/java/com/epam/controller/MainController.java (hunk 1 above)
Comment (ghost) [reply src/main/java/com/epam/controller/MainController.java]: fix it in all places in your PR

File: src/main/resources/schema.sql [hunk 16]
```diff
@@ -39,13 +39,13 @@ CREATE TABLE tourOffer
...
-  client_id       INT          NOT NULL,
-  tourOffer_id    INT          NOT NULL,
-  numberOfPersons INT          NOT NULL,
-  status          VARCHAR(100) NOT NULL,
-  discount_id     INT,
-  totalPrice      INT,
+  id             NUMERIC IDENTITY PRIMARY KEY,
+  client_id      INT          NOT NULL,
+  tourOffer_id   INT          NOT NULL,
+  numberOfPeople INT          NOT NULL,
```
Comment (ghost) [inline src/main/resources/schema.sql]: In DB we have some column naming conventions too. Please try to follow them, there is no camel case here

File: src/main/webapp/WEB-INF/jsp/testadmin.jsp [hunk 17]
```diff
@@ -1,9 +1,339 @@
-<%@ page contentType="text/html;charset=UTF-8"%>
+<%@ taglib prefix="spring" uri="http://www.springframework.org/tags" %>
+<%@ taglib prefix='c' uri='http://java.sun.com/jsp/jstl/core' %>
+<%@ page contentType="text/html;charset=UTF-8" %>
 <html>
 <head>
-    <title>Test page</title>
+    <title>Admin Page</title>
```
Comment (ghost) [inline src/main/webapp/WEB-INF/jsp/testadmin.jsp]: It is from another pull request, isnt it?

File: src/test/java/com/epam/repository/ReservationDAOTest.java [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+        when(resultSet.getInt("id")).thenReturn(testId);
+        when(resultSet.getInt("client_id")).thenReturn(testClientId);
+        when(resultSet.getInt("tourOffer_id")).thenReturn(testTourOfferId);
+        when(resultSet.getInt("numberOfPeople")).thenReturn(testNumberOfPeople);
+        when(resultSet.getString("status")).thenReturn(testStatus);
+        when(resultSet.getInt("discount_id")).thenReturn(testDiscountId);
+        when(resultSet.getInt("totalPrice")).thenReturn(testTotalPrice);
+        when(expectedReservation.getId()).thenReturn(testId);
+        Reservation actualReservation = reservationDAO.buildReservation(resultSet);
+        assertEquals(expectedReservation.getId(), actualReservation.getId());
```
Comment (ghost) [inline src/test/java/com/epam/repository/ReservationDAOTest.java:51]: If you check only id here, why you need all other stubs?

File: src/test/java/com/epam/repository/ReservationDAOTest.java [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Test(expected = SQLException.class)
+    public void buildPersonThrowsException() throws SQLException {
```
Comment (ghost) [inline src/test/java/com/epam/repository/ReservationDAOTest.java]: not only ThrowsException, ThrowsSqlException_the_short_reason

File: src/test/java/com/epam/repository/ReservationDAOTest.java [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+    private Integer testDiscountId = 1;
+    private Integer testTotalPrice = 10_000;
+
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void buildPersonPositiveCheck() throws SQLException {
```
Comment (ghost) [inline src/test/java/com/epam/repository/ReservationDAOTest.java]: without positiveCheck please

Comment (ghost) [inline src/test/java/com/epam/repository/ReservationDAOTest.java]: See the @SneakyThrows annotation for the exceptions

File: src/test/java/com/epam/service/ReservationServiceTest.java [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+        Integer testClientId = 1;
+        Integer testTourOfferId = 1;
+        Integer testNumberOfPeople = 1;
+        String testStatus = "paid";
+        Integer testDiscountId = 1;
+        Integer testTotalPrice = 10_000;
+        expectedReservation = new Reservation(testId, testClientId, testTourOfferId, testNumberOfPeople, testStatus, testDiscountId, testTotalPrice);
```
Comment (ghost) [inline src/test/java/com/epam/service/ReservationServiceTest.java]: you can use @Mock here

File: src/test/java/com/epam/service/ReservationServiceTest.java [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+        reservationService = new ReservationService(reservationDAO);
```
Comment (ghost) [inline src/test/java/com/epam/service/ReservationServiceTest.java]: and here

File: src/test/java/com/epam/repository/ReservationDAOTest.java (hunk 19 above)
Comment (ghost) [reply src/test/java/com/epam/repository/ReservationDAOTest.java]: please fix it everywhere

File: src/test/java/com/epam/repository/ReservationDAOTest.java (hunk 20 above)
Comment (ghost) [reply src/test/java/com/epam/repository/ReservationDAOTest.java]: fix it everywhere please

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/main/java/com/epam/model/Reservation.java (hunk 4 above)
Comment (goponenkoev) [reply src/main/java/com/epam/model/Reservation.java:8]: only @RequiredArgsConstructor https://projectlombok.org/features/Data

File: src/main/java/com/epam/model/Reservation.java (hunk 6 above)
Comment (goponenkoev) [reply src/main/java/com/epam/model/Reservation.java]: Good idea

File: src/main/java/com/epam/repository/ReservationDAO.java (hunk 14 above)
Comment (MaltsevaKsenia) [reply src/main/java/com/epam/repository/ReservationDAO.java]: can I create another task for that?

File: src/main/java/com/epam/controller/ReservationController.java (hunk 2 above)
Comment (MaltsevaKsenia) [reply src/main/java/com/epam/controller/ReservationController.java]: done

File: src/main/java/com/epam/model/Reservation.java (hunk 5 above)
Comment (MaltsevaKsenia) [reply src/main/java/com/epam/model/Reservation.java]: done

File: src/test/java/com/epam/repository/ReservationDAOTest.java (hunk 20 above)
Comment (MaltsevaKsenia) [reply src/test/java/com/epam/repository/ReservationDAOTest.java]: done

Comment (MaltsevaKsenia) [reply src/test/java/com/epam/repository/ReservationDAOTest.java]: done

File: src/test/java/com/epam/service/ReservationServiceTest.java (hunk 21 above)
Comment (MaltsevaKsenia) [reply src/test/java/com/epam/service/ReservationServiceTest.java]: done

File: src/test/java/com/epam/service/ReservationServiceTest.java (hunk 22 above)
Comment (MaltsevaKsenia) [reply src/test/java/com/epam/service/ReservationServiceTest.java]: done

File: src/main/java/com/epam/controller/ReservationController.java (hunk 3 above)
Comment (MaltsevaKsenia) [reply src/main/java/com/epam/controller/ReservationController.java]: I thought that we should use only get and post methods

File: src/main/java/com/epam/model/Reservation.java (hunk 6 above)
Comment (MaltsevaKsenia) [reply src/main/java/com/epam/model/Reservation.java]: done

File: src/main/java/com/epam/repository/ReservationDAO.java (hunk 8 above)
Comment (MaltsevaKsenia) [reply src/main/java/com/epam/repository/ReservationDAO.java]: done

File: src/main/java/com/epam/repository/ReservationDAO.java (hunk 9 above)
Comment (MaltsevaKsenia) [reply src/main/java/com/epam/repository/ReservationDAO.java]: done

File: src/main/java/com/epam/config/DBConfiguration.java (hunk 10 above)
Comment (MaltsevaKsenia) [reply src/main/java/com/epam/config/DBConfiguration.java]: done in another pr

File: src/main/java/com/epam/repository/ReservationDAO.java (hunk 12 above)
Comment (MaltsevaKsenia) [reply src/main/java/com/epam/repository/ReservationDAO.java]: done

File: src/main/java/com/epam/repository/ReservationDAO.java (hunk 13 above)
Comment (MaltsevaKsenia) [reply src/main/java/com/epam/repository/ReservationDAO.java]: done

File: src/main/resources/schema.sql (hunk 16 above)
Comment (MaltsevaKsenia) [reply src/main/resources/schema.sql]: done

File: src/test/java/com/epam/repository/ReservationDAOTest.java (hunk 18 above)
Comment (MaltsevaKsenia) [reply src/test/java/com/epam/repository/ReservationDAOTest.java:51]: done

File: src/main/java/com/epam/repository/ReservationDAO.java (hunk 14 above)
Comment (ghost) [reply src/main/java/com/epam/repository/ReservationDAO.java]: yes

File: src/main/java/com/epam/controller/ReservationController.java [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+    @GetMapping("/testadmin")
+    public String listReservations(ModelMap modelMap) {
+        List<Reservation> list = reservationService.listReservations();
+        modelMap.addAttribute("listReservation", list);
+        return "testadmin";
+    }
+
+    @GetMapping("/deleteReservation/{id}")
+    public String deleteReservation(@PathVariable int id) {
+        reservationService.removeReservation(id);
```
Comment (ghost) [inline src/main/java/com/epam/controller/ReservationController.java]: How do you will know that deleting was successful?

File: src/main/java/com/epam/model/Person.java [hunk 24]
```diff
@@ -7,7 +7,7 @@
 @Data
 public class Person {
     private Integer id;
-    private String name;
+    private String email;
```
Comment (ghost) [inline src/main/java/com/epam/model/Person.java]: Can you merge current dev in your branch and track, that this PR contains only reservation stuff changes?

File: src/main/java/com/epam/service/ReservationService.java [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+    private final ReservationDAO reservationDAO;
+
+
+    @Autowired
+    public ReservationService(ReservationDAO reservationDAO) {
+        this.reservationDAO = reservationDAO;
+    }
+
+
+    int addReservation(Reservation reservation) {
```
Comment (ghost) [inline src/main/java/com/epam/service/ReservationService.java:22]: Whats happen if i pass null here?

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/main/java/com/epam/service/ReservationService.java (hunk 15 above)
Comment (MaltsevaKsenia) [reply src/main/java/com/epam/service/ReservationService.java]: done

File: src/main/java/com/epam/controller/ReservationController.java (hunk 23 above)
Comment (MaltsevaKsenia) [reply src/main/java/com/epam/controller/ReservationController.java]: done

File: src/main/java/com/epam/service/ReservationService.java (hunk 25 above)
Comment (MaltsevaKsenia) [reply src/main/java/com/epam/service/ReservationService.java:22]: done?

File: src/main/java/com/epam/repository/ReservationDAO.java (hunk 14 above)
Comment (MaltsevaKsenia) [reply src/main/java/com/epam/repository/ReservationDAO.java]: done

File: src/main/java/com/epam/service/ReservationService.java (hunk 25 above)
Comment (ghost) [reply src/main/java/com/epam/service/ReservationService.java:22]: In this method - yes. How about other methods?

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/main/java/com/epam/service/ReservationService.java [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+
+    int getTourOfferById(Integer tourOfferId) {
+        if (tourOfferId != null) {
+            if (tourOfferId >= 0) {
+        return reservationDAO.getTourOfferById(tourOfferId);
+            } else {
+                throw new IllegalArgumentException("Id must be >= 0");
+            }
+        } else {
+            throw new RuntimeException("Tour offer with such id does not exist.");
```
Comment (ghost) [inline src/main/java/com/epam/service/ReservationService.java]: please, specify more certain exception here, for example the same IllegalArgument maybe

**Review by ghost — APPROVED** [review]

**Review by goponenkoev — APPROVED** [review]

**Review by grigiv — APPROVED** [review]

**Review by Cisfinitum — APPROVED** [review]
